cuda_phase-retrieval_class/test/test_6_cupy_reference_management/test_algo.hpp

```diff
@@ -0,0 +1,595 @@
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+#include <cufft.h>
+#include <cuComplex.h>
+#include <curand.h>
+#include <curand_kernel.h>
+#include <cmath>
+#include <cstdio>
+#include <time.h>
+#include <iostream>
+#include <complex>
+#include <string>
+#include <random>
+#if defined _MSC_VER
+#include <direct.h>
+#define GetCurrentDir _getcwd
+#else
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#define GetCurrentDir getcwd
+#endif
+
+#define PI 3.1415926535897932384626433
+#define CUDA_CHECK(call) {cudaError_t error = call; if(error!=cudaSuccess){printf("<%s>:%i ",__FILE__,__LINE__); printf("[CUDA] Error: %s\n", cudaGetErrorString(error));}}
+using namespace std;
+using namespace std::literals::complex_literals;
+namespace py = pybind11;
+
+void createDir(string dir);
+string get_current_dir();
+__device__ cufftDoubleComplex gpu_exp(cufftDoubleComplex arg);
```

Comment:
There is a lot of copied code from `src/phase_algo.hpp` in this test. Please remove the copy of the code from this file and include the original code. I think you need to create a new header file therefore.

Beside, it is also useless to use a copy a of the productive code, because you only tests the copy and not the actual code, which you want to use.

Reply:
this is done. please check